File: structures/C++/heap.cpp
#include<bits/stdc++.h>

using namespace std;

class Heap{
private:
    vector<int> heap;

    void heapifyUp(int index){
        if(index != 0){
            int parent = (index-1)/2;
            if(heap[parent]<heap[index]){
                swap(heap[parent],heap[index]);
                heapifyUp(parent);
            }
        }
    }

    void heapifyDown(int index){
        if(index<heap.size()){
            int left = 2*index+1, right = 2*index+2;
            vector< pair<int,int> > children;
            
            if(left<heap.size()){
                children.push_back(make_pair(heap[left],left));
            }

            if(right<heap.size()){
                children.push_back(make_pair(heap[right],right));
            }

            if(!children.empty()){
                int maxIndex = 0;
                for(int i=1;i<children.size();i++){
                    if(children[maxIndex].first<children[i].first){
                        maxIndex = i;
                    }
                }

                int toHeapify = children[maxIndex].second;
                if(heap[index]<heap[toHeapify]){
                    swap(heap[index],heap[toHeapify]);
                    heapifyDown(toHeapify);
                }
            }
        }
    }

public:
    Heap(){
        heap.clear();
    }

    void push(const int &value){
        heap.push_back(value);
        heapifyUp(heap.size()-1);
    }

    void pop(){
        if(heap.empty())
            return;

        heap.front() = heap.back();
        heap.pop_back();
        if(!heap.empty()){
            heapifyDown(0);
        }
    }

    int peek(){
        if(heap.empty()){
            return -1;
        }
        return heap.front();
    }

    void printHeap(){
        for(auto val : heap){
            cout<<val<<" ";
        }
        cout<<endl;
    }

    bool empty(){
        return heap.empty();
    }
};

int main(){

    Heap heap;
    heap.push(4);
    heap.push(1);
    heap.push(6);
    heap.push(2);
    heap.push(3);
    heap.push(10);
    while(!heap.empty()){
        cout<<heap.peek()<<" "; 
        heap.pop();  
    }
     
    return 0;
}

File: Kotlin/src/main/kotlin/org/algorithm/structures/Heap.kt
package org.algorithm.structures

class Heap<T:Comparable<T>>(val isMinHeap : Boolean = false){

    private val heap: MutableList<T> = mutableListOf<T>()

    private fun heapifyUp(index : Int){
        if(index>=0){
            val parent: Int = (index-1)/2
            if(heap[index]>heap[parent]){
                heap.swap(index,parent)
                heapifyUp(parent)
            }
        }
    }

    private fun heapifyDown(index : Int){
        if(index<heap.size){
            val left = 2*index+1
            val right = 2*index+2

            val children = mutableListOf<Pair<T,Int>>()

            if(left<heap.size){
                children.add(Pair(heap[left],left))
            }

            if(right<heap.size){
                children.add(Pair(heap[right],right))
            }

            if(!children.isEmpty()){
                var swapIndex = 0
                for(i in 1 until children.size){
                    if(children[swapIndex].first < children[i].first){
                        swapIndex = i
                    }
                }

                val toHeapify = children[swapIndex].second
                if(heap[toHeapify]>heap[index]){
                    heap.swap(index,toHeapify);
                    heapifyDown(toHeapify);
                }
            }
        }
    }

    fun printHeap(){
        println(heap)
    }

    fun push(value : T){
        heap.add(value)
        heapifyUp(heap.size-1)
    }

    fun peek() : T{
        return heap.first()
    }

    fun isEmpty() : Boolean{
        return heap.isEmpty()
    }

    fun pop(){
        if(!heap.isEmpty()){
            heap.removeAt(0)
            if(heap.isNotEmpty()){
                heapifyDown(0)
            }
        }
    }

    fun <T> MutableList<T>.swap(index1: Int, index2: Int) {
        val tmp = this[index1]
        this[index1] = this[index2]
        this[index2] = tmp
    }
}